sfge/src/test/java/com/salesforce/rules/usewithsharingondatabaseoperation/SharingPolicyComplexInheritanceTest.java

```diff
@@ -0,0 +1,354 @@
+package com.salesforce.rules.usewithsharingondatabaseoperation;
+
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * tests MyClass instantiating and calling a method with a database operation method in Two, which
+ * extends One. Also, a few misc. tests below.
+ */
+public class SharingPolicyComplexInheritanceTest extends UseWithSharingOnDatabaseOperationTest {
+
+    // spotless:off
+    private static final String ONE_SOURCE =
+        "public %s class One {\n"
+            + "void op() {\n"
+            + "}\n"
+        + "}\n";
+
+    private static final String TWO_SOURCE =
+        "public %s class Two extends One {\n"
+            + "void op(Account a) {\n"
+            + "   %s;"
+            + "}\n"
+        + "}\n";
+
+    private static final String MY_CLASS_SOURCE =
+        "public %s class " + MY_CLASS + " {\n"
+        + "void foo(Account a) {\n"
+        + "    Two t = new Two();\n"
+        + "    t.op(a);\n"
+        + "}\n"
+    + "}\n";
+    // spotless:on
+
+    private static final int SINK_LINE = 3;
+
+    // there are 4 * 4 * 4 = 64 possible combinations with three sharing policies
+
+    @MethodSource("provideNoViolationTests")
+    @ParameterizedTest
+    public void testInheritanceNoViolation(
+            String operation, String policyOne, String policyTwo, String policyThree) {
+        String[] sourceCode = {
+            String.format(ONE_SOURCE, policyOne),
+            String.format(TWO_SOURCE, policyTwo, operation),
+            String.format(MY_CLASS_SOURCE, policyThree)
+        };
+        assertNoViolation(RULE, sourceCode);
+    }
+
+    private static Stream<Arguments> provideNoViolationTests() {
+        Stream.Builder<Arguments> sb = Stream.builder();
+        Stream<Arguments> databaseOperations = provideSelectDatabaseOperations();
+
+        String[] noViolationSharingPolicyCombos = {
```

Comment:
Add comment explaining which entry applies to which class, and how those classes interact. Also, consider using 2-d array.